### ISSUE TITLE
Lodash fails to use class methods

### ISSUE THREAD

**Issue by sateffen** [lodash/lodash#1988]
Hey there,

I don't know if this is actually an usecase, but this code:

    'use strict';
    class MyClass {
        constructor(factor) {
            this.list = [1, 2, 3];
            this.factor = factor;
        }

        multiplyByFactor(n) {
            return n * this.factor;
        }

        getMultiplied() {
            return _.map(this.list, this.multiplyByFactor);
        }
    }

    const instance = new MyClass();
    console.log(instance.getMultiplied());

Will fail with *TypeError: Cannot read property 'factor' of undefined*. Even when passing `this` as third parameter to `_.map` (this is not documented anymore, so I guess it's not supported anymore).

In the previous version (3.10.0) of lodash this works fine, so in my opinion this is a huge regression, or is this intended?

Fiddle: https://jsfiddle.net/ho2L1g8k/1/

The fiddle works in up to date chrome, or copy the sourcecode for usage in nodejs (don't forget to load lodash, the fiddle loads version 4.3.0 from cdnjs.com)

Edit: Updated fiddle

**Comment by jfmengels** [comment]
In v4, the 3rd argument to set `this` has indeed been removed. You can pass `this.multiplyByFactor.bind(this)` to you `_.map` call and this should work.

Note: in your jsfiddle, you'll get `NaN` because you don't pass a factor when instanciating your `MyClass`.

**Comment by sateffen** [comment]
You're right, forgot the constructor parameter.

But is this actually intended, that lodash overwrites the context of the iteratee function? I don't see, why this would be a good idea.

I could rework my code base to actually bind everything everywhere, but this makes the code smelling in my opinion, and I want to understand, why this change was made.

**Comment by jfmengels** [comment]
This is intended. I'm not sure about the exact reasons, but I think it's in an attempt to simplify things.

About the fact that binding everything is a code smell, I agree. There are two things you can do:
- Use ES2015 arrow functions for implicit this binding where possible
- Try to use `this` as little as possible. For instance in this case, the `multiplyByFactor` can easily be extracted from the class and made a pure function that takes two arguments `n` and `factor`. You can then just pass the value of factor to your function. This would be the best option IMO

**Comment by demoneaux** [comment]
@jfmengels is right, you can use `_.bind`, `Function#bind` or arrow functions too. See https://github.com/lodash/lodash/issues/1878#issuecomment-174671246 for a brief rationale.

**Comment by jfmengels** [comment]
I've added the rationale in the changelog (wiki).

**Comment by jdalton** [comment]
@sateffen 
> But is this actually intended, that lodash overwrites the context of the iteratee function? I don't see, why this would be a good idea.

Lodash isn't overwriting anything. In JavaScript when you provide a callback value like `this.multiplyByFactor` to a function it will be executed by default with a `this` binding of the global object or `undefined` (if in strict mode). It's disassociated from the binding of `this` which is why you need to use `Function#bind`, or `_.bind` or arrow functions.

**Comment by sateffen** [comment]
Ahhh, I learned something today. I just checked it myself, because I didn't beleaved you, but you're absolutly right. The callback gets executed in the execution context, not code context. Hmm, classes suck more and more somehow^^

Big thanks @jdalton , now I get it, why it is like it is. I checked my code, I used `_.map(this.list, this.func, this)` everywhere, without thinking about it, but now I know why it was intentionally right.
